src/data/equiv/algebra.lean

```diff
@@ -355,4 +380,112 @@ instance symm.is_ring_hom {e : α ≃r β} : is_ring_hom e.to_equiv.symm := hom
 @[simp] lemma to_equiv_symm_apply (e : α ≃r β) (x : β) :
   e.symm.to_equiv x = e.to_equiv.symm x := rfl
 
+/-- Two ring isomorphisms agree if they are defined by the
+    same underlying function. -/
+@[extensionality] lemma ext {R S : Type*} [ring R] [ring S]
+  {f g : ring_equiv R S} (h : f.to_fun = g.to_fun) : f = g :=
+begin
+  have h₁ := @equiv.eq_of_to_fun_eq _ _ f.to_equiv g.to_equiv h,
+  cases f, cases g, congr,
+  { exact h },
+  { exact congr_arg equiv.inv_fun h₁ }
+end
+
 end ring_equiv
+
+/-
+We define the type of automorphisms on groups, additive monoids,
+additive groups and rings using `mul_equiv`, `add_equiv` and `ring_equiv`.
+In each case this type also forms a group.
+-/
+
+namespace monoid
+
+def aut (γ : Type u) [has_mul γ] := mul_equiv γ γ
+
+/--
+The group operation on automorphisms of a monoid is defined by
+λ g h, mul_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (γ : Type u) [has_mul γ] : group (aut γ) :=
+{ mul := λ g h, mul_equiv.trans h g,
+  one := mul_equiv.refl γ,
+  inv := mul_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+end monoid
+
+namespace group
+
+-- A group homomorphism is a monoid homomorphism between groups.
+def aut (γ : Type u) [group γ] := mul_equiv γ γ
+
+instance aut_group (γ : Type u) [group γ] : group (aut γ) := monoid.aut_group γ
+
+end group
+
+namespace add_monoid
+
+def aut (α : Type u) [has_add α] := add_equiv α α
+
+/--
+The group operation on automorphisms of an additive monoid is defined by
+λ g h, add_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (α : Type u) [has_add α] : group (aut α) :=
+{ mul := λ g h, add_equiv.trans h g,
+  one := add_equiv.refl α,
+  inv := add_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+end add_monoid
+
+namespace add_group
+
+-- An additive group homomorphism is an additive monoid homomorphism between groups.
+def aut (γ : Type u) [add_group γ] := add_equiv γ γ
+
+instance aut_group (γ : Type u) [add_group γ] : group (aut γ) := add_monoid.aut_group γ
+
+end add_group
+
+namespace ring
```

Comment:
As above, if `ring_equiv` only needs `semiring`, please weaken the assumptions in this section.

Reply:
I'm not going to address this until we decide how to handle all the duplicated `.aut` definitions.

Reply:
About duplicate definitions: I'd just introduce
```lean
@[to_additive] def mul_aut {α : Type u} [has_mul α] := α ≃* α

def ring_aut {α : Type u} [has_mul α] [has_add α] := α ≃+* α
```

This way it is similar to `mul_equiv` etc.